Core/include/Acts/TrackFinding/ComposableTrackStateCreator.hpp

```diff
@@ -0,0 +1,415 @@
+// This file is part of the ACTS project.
+//
+// Copyright (C) 2016 CERN for the benefit of the ACTS project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+// for definitions of Calibrator, MeasurementSelector
+#include "Acts/TrackFitting/KalmanFitter.hpp"
+
+namespace Acts {
+/// to be processed by the CKF
+template <typename source_link_iterator_t>
+using SourceLinkAccessorDelegate =
+    Delegate<std::pair<source_link_iterator_t, source_link_iterator_t>(
+        const Surface&)>;
+
+/// expected max number of track states that are expected to be added by
+/// stateCandidateCreator
+/// @note if the number of states exceeds this number dynamic memory allocation will occur.
+///       the number is chosen to yield a container size of 64 bytes.
+static constexpr std::size_t s_maxBranchesPerSurface = 10;
+
+namespace CkfTypes {
+
+template <typename T>
+using BranchVector = boost::container::small_vector<T, s_maxBranchesPerSurface>;
+
+}  // namespace CkfTypes
+
+namespace detail {
+// helper method to deduce the return type and arguments of the member function
+template <auto memberfunc_ptr, typename return_t, typename instance_t,
+          typename... Args>
+inline auto makeDelegate(
+    [[maybe_unused]] return_t (instance_t::*memberfunc)(Args...) const,
+    const instance_t& instance) {
+  Delegate<return_t(Args...)> delegate;
+  delegate.template connect<memberfunc_ptr>(&instance);
+  return delegate;
+}
+// helper method to deduce the return type and arguments of the function
+template <auto func_ptr, typename return_t, typename... Args>
+inline auto makeDelegate([[maybe_unused]] return_t (*func)(Args...)) {
+  Delegate<return_t(Args...)> delegate;
+  delegate.template connect<func_ptr>();
+  return delegate;
+}
+}  // namespace detail
+
+/// @brief convenience method to create and connect a delegate.
+/// @tparam memberfunc_ptr a pointer to a member function of the class instance_t
+/// @param instance an instance which outlives this delegate.
+template <auto memberfunc_ptr, typename instance_t>
+inline auto makeDelegate(const instance_t& instance) {
+  return detail::makeDelegate<memberfunc_ptr>(memberfunc_ptr, instance);
+}
+
+/// @brief convenience method to create and connect a delegate.
+/// @tparam func_ptr a pointer to a function
+template <auto func_ptr>
+inline auto makeDelegate() {
+  return detail::makeDelegate<func_ptr>(func_ptr);
+}
```

Comment:
This seems a bit generic for a header called `ComposableTrackStateCreator.hpp`. Either this should be a utility placed in the delegate header or dropped in favor of doing this by hand.

I think the deduction should also work without a function wrapper?

Reply:
I removed these helper functions, although I think that they would have been useful